### PR TITLE
NAS-125228 / 24.04 / Improvements to query grammar

### PR DESCRIPTION
**Pull request by AlexKarpov98**
Testing: go to http://localhost:4200/audit

Read the ticket & see that tests are passing:
In Audit → Advanced Search improved query language to:

- Allow non-ascii non-quoted property names. E.g. currently "Имя"="Bob" is supported, but Имя="Bob" isn’t.
- Do not allow unterminated strings in values. E.g. currently Username = "Bob is a valid string. It probably shouldn’t be.

___ 

Check console logs and see that it works as described in ticket

### CODE REVIEW

File: src/app/modules/search-input/services/query-parser/tests/property-names.spec.ts [hunk 1]
```diff
@@ -100,5 +106,25 @@ describe('QueryParserService - property names', () => {
...
+      const noQuotesAndSpace = service.parseQuery('Я є = "Bob"');
+      expect(noQuotesAndSpace.hasErrors).toBe(true);
+      expect(noQuotesAndSpace.errors).toHaveLength(1);
+      expect(noQuotesAndSpace.errors[0]).toBeInstanceOf(QuerySyntaxError);
+
+      const noQuotesChinese = service.parseQuery('姓名 = "Bob"');
+      expect(noQuotesChinese.hasErrors).toBe(false);
+      expect(noQuotesChinese.tree).toMatchObject({ 'property': '姓名' });
+
+      const noQuotesChineseAndSpace = service.parseQuery('姓名 姓名 = "Bob"');
```
[the remark this answers is not in the document]
Comment (AlexKarpov98) [reply src/app/modules/search-input/services/query-parser/tests/property-names.spec.ts:124]: sure, currently we don't allow spaces in unquoted strings

[the remark this answers is not in the document]
Comment (undsoft) [reply src/app/modules/search-input/services/query-parser/tests/property-names.spec.ts:124]: Ah, you are right!

**Comment by AlexKarpov98** [comment]
@undsoft  - now it works as expected, can you check, please?

**Review by denysbutenko — COMMENTED** [review]
When I type:

- `Имя="Bob"` - in console logs I've seen `[["Имя","=","Bob"]]`
- `Username = "Bob` - no console logs is invoked. If adding quotes to the end, then it shows `[["Username","=","Bob"]]` in the logs.

Should I see search results on the table or it is not implemented yet?

**Comment by AlexKarpov98** [comment]
> When I type:
> 
> * `Имя="Bob"` - in console logs I've seen `[["Имя","=","Bob"]]`
> * `Username = "Bob` - no console logs is invoked. If adding quotes to the end, then it shows `[["Username","=","Bob"]]` in the logs.
> 
> Should I see search results on the table or it is not implemented yet?

search results are not implemented yet, 2 scenarios you wrote - are correct.

File: src/app/modules/search-input/services/query-parser/query.grammar [hunk 2]
```diff
@@ -4,14 +4,15 @@
 // yarn run compile-grammar
 
 @tokens {
-  space { @whitespace+ }
+  space { " " | "\\t" | "\\n" | "\\r" }
```
Comment (undsoft) [inline src/app/modules/search-input/services/query-parser/query.grammar:7]: What's the reason for this change?

**Review by undsoft — CHANGES_REQUESTED** [review]
This condition doesn't work for me now:
```
Username = "Bob"
AND Username > "Sarah"
```

File: src/app/modules/search-input/services/query-parser/query.grammar (hunk 2 above)
Comment (AlexKarpov98) [reply src/app/modules/search-input/services/query-parser/query.grammar:7]: @undsoft  the reason is that using `@whitespace+` will result in compilation errors:
```
Overlapping tokens UnquotedString and space used in same context (example: "") 
```

We have a broad list of available symbols in here, that leads to lezer conflicts with `@whitespace`:
UnquotedString { @asciiLetter+ | $[a-zA-Z\u0080-\uFFFF]+ }
Extracting white spaces from this regex is not possible due to lezer's limitations, at least I tried different ways to do it and no success.

So I decided to override it and it works now.

**Review by undsoft — APPROVED** [review]
Looks good, thanks.
